controller/runner.go

```diff
@@ -0,0 +1,186 @@
+	r = &ServerRunner{
+		log: logp.L(),
+		cfg: cfg,
+	}
+	// in case of an initialization error we must clean up all created resources
+	defer func() {
+		if err != nil {
+			// this will account for partial initialization in case of an error, so there are no leaks
+			r.Close()
+		}
+	}()
+
+	r.log.Debug("initializing the queue...")
+	r.queue, err = queue.New(cfg.Queue)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create queue: %w", err)
+	}
+	r.log.Debug("queue was initialized.")
+
+	r.log.Debug("initializing monitoring ...")
+	r.monitoring, err = monitoring.NewFromConfig(cfg.Monitor, r.queue)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing output monitor: %w", err)
+	}
+	r.monitoring.Watch()
+	r.log.Debug("monitoring is ready.")
+
+	r.log.Debug("initializing the output...")
+	// TODO replace with the real output based on the config, Console is hard-coded for now
+	r.out = output.NewConsole(r.queue)
+	r.out.Start()
+	r.log.Debug("output was initialized.")
+
+	r.log.Debug("initializing the gRPC server...")
+	var opts []grpc.ServerOption
+	if cfg.Server.TLS {
+		creds, err := credentials.NewServerTLSFromFile(cfg.Server.Cert, cfg.Server.Key)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate credentials %w", err)
+		}
+		opts = []grpc.ServerOption{grpc.Creds(creds)}
+	}
+	r.server = grpc.NewServer(opts...)
+	r.shipper, err = server.NewShipperServer(cfg.Server, r.queue)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialise the server: %w", err)
+	}
+	pb.RegisterProducerServer(r.server, r.shipper)
+	r.log.Debug("gRPC server initialized.")
+
+	return r, nil
+}
+
+// Start runs the shipper server according to the set configuration.
+// The server stops after calling `Close`, this function blocks until then.
+func (r *ServerRunner) Start() (err error) {
+	r.mu.Lock()
+	if r.server == nil {
+		r.mu.Unlock()
+		return fmt.Errorf("failed to start a server runner that was previously closed")
+	}
+
+	addr := fmt.Sprintf("localhost:%d", r.cfg.Server.Port)
+	lis, err := net.Listen("tcp", addr)
+	if err != nil {
+		r.mu.Unlock()
+		return fmt.Errorf("failed to listen on %s: %w", addr, err)
+	}
+
+	// Testing that the server is running and only then unlock the mutex.
+	// Otherwise if `Close` is called at the same time as `Start` it causes race condition.
+	go func() {
+		defer r.mu.Unlock()
+		con, err := net.Dial("tcp", addr)
+		if err != nil {
+			r.log.Errorf("failed to test connection with the gRPC server on %s", addr)
+			return
+		}
+		defer con.Close()
+		r.log.Debugf("gRPC server is ready and is listening on %s", addr)
+	}()
+
+	return r.server.Serve(lis)
+}
+
+// Close shuts the whole shipper server down. Can be called only once, following calls are noop.
+func (r *ServerRunner) Close() (err error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	// initialization can fail on each step, so it's possible that the runner
+	// is partially initialized and we have to account for that.
+	r.once.Do(func() {
```

Comment:
Why does this need to be in a once block?

Reply:
You're right, after I added `nil` checks there is no need anymore, I'll remove it.

Reply:
After a second thought, I think we need it in case there will be several concurrent calls of `Close` by the control protocol. We don't want a race condition with `nil` checks, I remember it now.